src/D2L.Security.OAuth2/Keys/KeyManagementService.cs

```diff
@@ -0,0 +1,207 @@
+
+			var keyId = Guid.NewGuid();
+			var keyIdStr = keyId.ToString();
+
+			using var csp = new RSACryptoServiceProvider {
+				PersistKeyInCsp = false
+			};
+
+			var pub = csp.ExportParameters(
+				includePrivateParameters: false
+			);
+
+			var priv = csp.ExportCspBlob(
+				includePrivateParameters: true
+			);
+
+			var expiresAt = now + m_config.KeyLifetime;
+
+			await Task.WhenAll(
+				m_publicKeys.SaveAsync(
+					keyId,
+					new RsaJsonWebKey(
+						keyIdStr,
+						expiresAt,
+						pub
+					)
+				),
+				m_privateKeys.SaveAsync(
+					new PrivateKeyData(
+						id: keyIdStr,
+						kind: PrivateKeyData.KeyKinds.Rsa,
+						data: priv,
+						createdAt: now,
+						notBefore: now + m_config.KeyTimeUntilUse,
+						expiresAt: expiresAt
+					)
+				)
+			).ConfigureAwait( false );
+		}
+
+		private async Task RefreshKeyAsync(
+			D2LSecurityToken current,
+			DateTimeOffset now
+		) {
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( continueOnCapturedContext: false );
+
+			var best = ChooseKey( keys, now ).Ref();
+
+			var loser = Interlocked.CompareExchange(
+				ref m_current,
+				best,
+				comparand: current
+			);
+
+			loser?.Dispose();
+		}
+
+		private D2LSecurityToken ChooseKey(
+			IEnumerable<PrivateKeyData> keys,
+			DateTimeOffset now
+		) {
+			PrivateKeyData candidate = null;
+
+			foreach( var key in keys ) {
+				// Ignore unsupported key types
+				if( key.Kind != PrivateKeyData.KeyKinds.Rsa ) {
+					continue;
+				}
+
+				// Use any non-expired key if it's the only one
+				if( candidate == null ) {
+					candidate = key;
+					continue;
+				}
+
+				if( candidate.NotBefore > now && key.NotBefore <= now ) {
+					// If we can switch to a key past its NotBefore, do that.
+					candidate = key;
+				} else if( candidate.NotBefore > now && candidate.CreatedAt > key.CreatedAt ) {
+					// When comparing two keys that are not past their respective
+					// NotBefore points, prefer the oldest one.
+					candidate = key;
+				} else if ( candidate.NotBefore <= now && key.NotBefore <= now && candidate.ExpiresAt < key.ExpiresAt ) {
+					// If we have two keys past their NotBefore date, pick the
+					// one that expires the latest.
+					candidate = key;
+				}
+			}
+
+			Func<Tuple<AsymmetricSecurityKey, IDisposable>> keyFactory = candidate.Kind switch {
+				PrivateKeyData.KeyKinds.Rsa => () => {
+					var csp = new RSACryptoServiceProvider {
+						PersistKeyInCsp = false
+					};
+
+					csp.ImportCspBlob( candidate.Data );
+
+					var key = new RsaSecurityKey( csp );
+
+					return new(key, csp);
+				},
+				_ => throw new NotImplementedException()
+			};
+
+			return new D2LSecurityToken(
+				id: candidate.Id,
+				validFrom: candidate.CreatedAt,
+				validTo: candidate.ExpiresAt,
+				keyFactory
+			);
+		}
+
+		private readonly object m_disposeLock = new();
+
+		public void Dispose() {
+			lock( m_disposeLock ) {
+				m_current?.Dispose();
+			}
+		}

```

Comment:
If we're disposable, should have disposed checks in methods I think?

Reply:
I believe the spirit (but not the law) is that you shouldn't use an object after it's disposed, other than possibly calling Dispose.
Sometimes there are `if( disposed )` checks within `Dispose` impls to support it being called twice, but here we're fine.

I'm not really sure what I was thinking with this lock though. By this logic `Dispose` needs to be threadsafe but the lock wasn't necessary for that here. So I think I'll go the other direction and remove it.